File: src/main/java/at/yawk/fiction/impl/fimfiction/TagsParser.java
package at.yawk.fiction.impl.fimfiction;

import at.yawk.fiction.impl.PageParser;
import java.net.URI;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;
import org.jsoup.nodes.Element;

/**
 * @author yawkat
 */
class TagsParser extends PageParser<List<FimTag>> {
    @Override
    protected List<FimTag> create() {
        return new ArrayList<>();
    }

    @Override
    protected void parse(Element root, List<FimTag> target) throws Exception {
        for (Element element : root.select("ul.tags-dropdown").first().select(">li")) {
            FimTag tag = new FimTag();
            tag.setName(element.attr("data-name"));
            tag.setId(element.attr("data-tag"));

            Element icon = element.select(".icon img").first();
            if (icon != null) {
                tag.setIcon(new URL(icon.absUrl("src")));
            }
            target.add(tag);
        }
    }
}

File: src/main/kotlin/at/yawk/fiction/impl/fimfiction/TagsParser.kt
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package at.yawk.fiction.impl.fimfiction

import at.yawk.fiction.impl.PageParser
import org.jsoup.nodes.Element
import java.net.URL

/**
 * @author yawkat
 */
internal object TagsParser : PageParser<List<FimTag>>() {
    override fun parse(root: Element): List<FimTag> {
        return root.select("ul.tags-dropdown").first().select(">li").map {
            val iconElement = it.select(".icon img").first()
            FimTag(
                    name = it.attr("data-name"),
                    id = it.attr("data-tag"),
                    icon = if (iconElement == null) null else URL(iconElement.absUrl("src"))
            )
        }
    }
}
